Added JSON support for reading rules and selectors from pipeline

ArmaanMcleod opened the pull request:
## PR Summary

Fix #857

Added JSON support for reading rules and selectors. 

Also did some refactoring and cleanup with some previous baseline code. 

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x] Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

BernieWhite commented on src/PSRule/Common/JsonConverters.cs:
```diff
@@ -400,70 +401,75 @@ private IResource MapResource(JsonReader reader, JsonSerializer serializer)
+                    var commentLine = reader.Value.ToString().TrimStart();
 
-                        if (commentLine.StartsWith("Synopsis: "))
+                    if (commentLine.StartsWith("Synopsis: "))
```
Maybe add a length check here as well.

```suggestion
                    if (commentLine.Length > 10 && commentLine.StartsWith("Synopsis: "))
```

BernieWhite reviewed (changes requested):
@ArmaanMcleod Looking good. A few minor changes suggested. Otherwise all good.

I did notice there was a problem with synopsis reporting but this was an existing problem that affected YAML resources as well. I added a commit to resolve this. Please double check.

ArmaanMcleod commented:
> @ArmaanMcleod Looking good. A few minor changes suggested. Otherwise all good.
> 
> I did notice there was a problem with synopsis reporting but this was an existing problem that affected YAML resources as well. I added a commit to resolve this. Please double check.

@BernieWhite Thanks for the feedback and also fixing the synopsis issue, love the new changes 🙂. 

Also fixed te docs a bit to emphasis using `.jsonc`. Let me know if you like these changes or would like to tweak it more.

BernieWhite commented on docs/concepts/PSRule/en-US/about_PSRule_Selectors.md:
````diff
@@ -137,6 +155,67 @@ spec:
     - 'Value2'
 ```
 
+### Example Selectors.Rule.json
````
```suggestion
### Example Selectors.Rule.jsonc
```

BernieWhite commented on docs/concepts/PSRule/en-US/about_PSRule_Selectors.md:
````diff
@@ -137,6 +155,67 @@ spec:
+### Example Selectors.Rule.json
+
+```json
+// Example Selectors.Rule.json
````
```suggestion
// Example Selectors.Rule.jsonc
```

BernieWhite commented on docs/concepts/PSRule/en-US/about_PSRule_Selectors.md:
```diff
@@ -73,10 +73,12 @@ If one or more selector pre-conditions are used, they are evaluated before type
+In either case, define a selector within a file ending with the `.Rule.yaml` or `.Rule.jsonc` extension.
 A selector can be defined side-by-side with other resources such as baselines or module configurations.
 
+Use `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```
```suggestion
Selectors can also be defined within `.json` files.
We recommend using `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```

BernieWhite commented on docs/concepts/PSRule/en-US/about_PSRule_Rules.md:
```diff
@@ -156,6 +180,65 @@ spec:
+
+Within the `Rule` resource, the `condition` property specifies conditions to pass or fail the rule.
+
+Use the `.jsonc` extension to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```
```suggestion
Rules can also be defined within `.json` files.
We recommend using `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```

BernieWhite reviewed (approved):
@ArmaanMcleod Nice! A few minor suggestions. Otherwise all good to merge.
